resourseleak/TestShadowMode2.java

```diff
@@ -0,0 +1,104 @@
+package resourceleak;
+
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.text.DateFormat;
+import java.util.Date;
+import java.util.Map;
+
+public class TestShadowMode2 {
+
+
+
+    public void testShadowMode10(Map<String, String> entry) {
+        if (entry.isEmpty()) {
+            return;
+        }
+
+        TSDMetrics tsdMetric = new TSDMetrics(t3MetricsFactory);
+        // set up marketplace
+        tsdMetric.addProperty("Marketplace", location);
+        int blackwatchBGPEstablishedCounter = 0;
+        for (String counterName : entry.keySet()) {
+            try {
+                long counterValue = Long.valueOf(entry.get(counterName));
+                // sum all BGP state counter together
+                if (counterName.contains(BGP_STATE_COUNTER_NAME)) {
+                    blackwatchBGPEstablishedCounter += counterValue;
+                }
+                tsdMetric.addCount(counterName, counterValue);
+            } catch (NumberFormatException ex) {
+                LOG.warn("unexpected counter value : " + entry.get(counterName));
+            }
+        }
+
+        // add counter for All BGP state counter
+        tsdMetric.addCount(BGP_STATE_COUNTER_NAME, blackwatchBGPEstablishedCounter);
+
+        // add counter for number of active host metric
+        // As each host maps 2 * portCount (number of ports in a host) established BGP session,
+        // so number of active host = (number of established BGP session) / (2 * portCountPerHost)
+        if (portsCount != 0) {
+            tsdMetric.addCount(
+                    BlackWatchCommon.ACTIVE_HOST_METRIC_NAME,
+                    (float) blackwatchBGPEstablishedCounter / (2 * portsCount));
+        }
+
+        tsdMetric.end();
+    }
+
+
+    public static void testShadowMode11(String[] args) throws IOException {
+        if (args.length != 1) {
+            System.err.println("USAGE: TraceFormatter trace_file");
+            System.exit(2);
+        }
+        FileChannel fc = new FileInputStream(args[0]).getChannel();
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *fc*. Currently, there are execution paths that do not contain closure statements. Either a) close *fc* in a try-finally block or b) close the resource by declaring *fc* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).